File: 2018/day02/day02-pt1.py
import os
import sys
from typing import Dict
from collections import Counter

directory: str = os.path.dirname(os.path.abspath(__file__))
filename: str = os.path.join(directory, 'input.txt')
two_count: int = 0
three_count: int = 0

with open(filename, 'r') as fp:
    for id_str in fp:
        char_count: Dict[str, int] = Counter(id_str)

        if 2 in char_count.values():
            two_count += 1

        if 3 in char_count.values():
            three_count += 1

print(f'checksum is {two_count * three_count}!')

File: 2018/day02/day02-pt1.go
package main

import (
	"fmt"
	"io/ioutil"
	"path/filepath"
	"strings"
)

func getInput() []string {
	filename, err := filepath.Abs("input.txt")
	data, err := ioutil.ReadFile(filename)
	if err != nil {
		panic(err)
	}

	idArray := strings.Split(string(data), "\n")

	return idArray
}

func inMap(num int, array map[string]int) bool {
	// inSlice checks to see if a value matches
	// and item in a slice.
	for _, value := range array {
		if num == value {
			return true
		}
	}

	return false
}

func countChars(id string) map[string]int {
	// countChars makes a map of the number of
	// the same like characters in each string
	count := make(map[string]int)

	for _, char := range id {
		char := string(char)

		if _, ok := count[char]; ok {
			count[char]++
		} else {
			count[char] = 1
		}
	}

	return count
}

// Main function
func main() {
	data := getInput()
	twoCount := 0
	threeCount := 0

	for _, id := range data {
		charCount := countChars(id)

		if inMap(2, charCount) {
			twoCount++
		}

		if inMap(3, charCount) {
			threeCount++
		}
	}

	fmt.Printf("The checksum is: %d\n", twoCount*threeCount)

}
